programs/quarter_board/communication.c

```diff
@@ -0,0 +1,163 @@
+            state = COMMAND_CHAR_0_READ; // communication reset - \r \n
+            sdPut(&SD1, received_char);
+            char_counter = 0;
+            command = 0;
+        } else if((received_char >= '0' && received_char <= '9') || \
+                  (received_char >= 'A' && received_char <= 'F')) {
+            
+            received_char = char_to_value(received_char);
+
+            if(state <= COMMAND_END) {
+                command |= (received_char << (state * CHAR_BIT_LENGHT));
+
+                sdPut(&SD1, value_to_char(received_char));
+
+                state++;
+
+                if(state == (COMMAND_END + 1)) {
+                    //commmand_received
+                    for(i = 0; i < (quantity_of_slaves * MAX_DATA_CHAR_LENGHT); i++) {
+                        operation_buffer[i] = NO_OPERATION;
+                        data_buffer[i] = 0;
+                    }
+
+                    data_lenght = (command >> (OPERATION_BIT_LENGHT + RW_BIT));
+                    op = (((command >> OPERATION_BIT_LENGHT) & 1) + 1);
+
+                    msgp -> operation = command;
+                    
+                    char_counter = 0;
+
+                    if(op == SEND_DATA_TO_MASTER) {
+                        msgp = (MESSAGE *)chMsgSend(msgThread, (msg_t)msgp);
+                        
+                        for(i = 0; i < data_lenght; i++) {
+                            data_buffer[i + (data_lenght * my_address)] = msgp -> data[i];
+                            operation_buffer[i + (data_lenght * my_address)] = SEND_DATA_TO_MASTER;
+                        }
+                    }
+
+                    if(op == PROCESS_DATA_FROM_MASTER) {
+                        state = PROCESSING_STATE;
+                        
+                        for(i = 0; i < data_lenght; i++) {
+                            operation_buffer[i + (data_lenght * my_address)] = PROCESS_DATA_FROM_MASTER;
+                        }
+                    }
+                }
+            } else {
+                if(state == SENDING_STATE || state == PROCESSING_STATE) {
+                    if(operation_buffer[char_counter] == NO_OPERATION) {
+                        sdPut(&SD1, value_to_char(received_char));
+                    } else if(operation_buffer[char_counter] == PROCESS_DATA_FROM_MASTER) {
+                        data_buffer[char_counter] = received_char;
+                        sdPut(&SD1, value_to_char(received_char));
+                    } else if(operation_buffer[char_counter] == SEND_DATA_TO_MASTER) {
+                        sdPut(&SD1, value_to_char(data_buffer[char_counter]));
+                    }
+
+                    char_counter++;
+ 
+                    if(char_counter == (quantity_of_slaves * data_lenght)) {
+                        if(state == PROCESSING_STATE) {
+                            for(i = 0; i < data_lenght; i++) {
+                                msgp -> data[i] = data_buffer[i + (data_lenght * my_address)];

```

Comment:
The `*` operator has higher priority than `+` operator, so the parentheses are not required. Please remove them.